{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import psycopg2\n",
    "import pandas as pd"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "os.chdir(\"/mnt/data0/proj_osgeo/data\")\n",
    "db_config=open('./db_config')\n",
    "dbHOST=db_config.readline().split('\\\"')[1]\n",
    "dbUSER=db_config.readline().split('\\\"')[1]\n",
    "dbPASS=db_config.readline().split('\\\"')[1]\n",
    "dbDB=db_config.readline().split('\\\"')[1]\n",
    "db_config.close()\n",
    "global db \n",
    "db = psycopg2.connect(host=dbHOST,user=dbUSER,password=dbPASS,database=dbDB)\n",
    "db.commit()\n",
    "cursor = db.cursor()"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "def save_one_proj(proj):\n",
    "    sql = \"\"\"INSERT INTO project(proj_id, proj_name, git_repo)\n",
    "                                        values('%s', '%s', '%s')\"\"\" % (\n",
    "        proj['id'], proj['title'], proj['source'])\n",
    "\n",
    "    try:\n",
    "        db.commit()\n",
    "        cursor.execute(sql)\n",
    "        db.commit()\n",
    "        print(\"%s : Insert successfully\" % proj['id'])\n",
    "    except Exception as e:\n",
    "        try:\n",
    "            db.commit()\n",
    "            print(\"Exception: %s\" % e)\n",
    "            sql_update = \"\"\"UPDATE project SET proj_name='%s', git_repo='%s' WHERE proj_id='%s'\"\"\" % (\n",
    "                proj['title'], proj['source'], proj['id'])\n",
    "            cursor.execute(sql_update)\n",
    "            db.commit()\n",
    "            print(\"%s : Update successfully\" % proj['id'])\n",
    "        except Exception as e:\n",
    "            logging.error(\"update error:%s\" % e)\n",
    "            db.rollback()\n",
    "            db.commit()"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "df = pd.read_csv(\"../gitrepository.csv\")"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "url = \"https://github.com/geopython/pywps\""
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "str(df.loc[df[\"Repos\"] == \"https://github.com/geopython/pywps\"][\"Projects\"].values).lower()"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "\"['pywps']\""
      ]
     },
     "metadata": {},
     "execution_count": 6
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "proj = {}\n",
    "for repo in df.itertuples():\n",
    "    #print(repo.Projects)\n",
    "    proj[\"id\"] = repo.Projects.lower()\n",
    "    proj[\"source\"] = repo.Repos\n",
    "    proj['title'] = repo.Projects\n",
    "    #print(proj)\n",
    "    save_one_proj(proj)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geonode) already exists.\n",
      "\n",
      "geonode : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geonode) already exists.\n",
      "\n",
      "geonode : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geonode) already exists.\n",
      "\n",
      "geonode : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geonetwork) already exists.\n",
      "\n",
      "geonetwork : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(pycsw) already exists.\n",
      "\n",
      "pycsw : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geonode) already exists.\n",
      "\n",
      "geonode : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(osgeolive) already exists.\n",
      "\n",
      "osgeolive : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(mapserver) already exists.\n",
      "\n",
      "mapserver : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(deegree) already exists.\n",
      "\n",
      "deegree : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(openlayers) already exists.\n",
      "\n",
      "openlayers : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geomoose) already exists.\n",
      "\n",
      "geomoose : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(mapbender) already exists.\n",
      "\n",
      "mapbender : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(pywps) already exists.\n",
      "\n",
      "pywps : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geoserver) already exists.\n",
      "\n",
      "geoserver : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(proj) already exists.\n",
      "\n",
      "proj : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotools) already exists.\n",
      "\n",
      "geotools : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geomoose) already exists.\n",
      "\n",
      "geomoose : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(gdal/ogr) already exists.\n",
      "\n",
      "gdal/ogr : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geoext) already exists.\n",
      "\n",
      "geoext : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(pygeoapi) already exists.\n",
      "\n",
      "pygeoapi : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(gc2/vidi) already exists.\n",
      "\n",
      "gc2/vidi : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geowebcache) already exists.\n",
      "\n",
      "geowebcache : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geomajas) already exists.\n",
      "\n",
      "geomajas : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(istsos) already exists.\n",
      "\n",
      "istsos : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geostyler) already exists.\n",
      "\n",
      "geostyler : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(mdal) already exists.\n",
      "\n",
      "mdal : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(actinia) already exists.\n",
      "\n",
      "actinia : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(owslib) already exists.\n",
      "\n",
      "owslib : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(ossim) already exists.\n",
      "\n",
      "ossim : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(pgrouting) already exists.\n",
      "\n",
      "pgrouting : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(loader) already exists.\n",
      "\n",
      "loader : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geohealthcheck) already exists.\n",
      "\n",
      "geohealthcheck : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(portable-gis) already exists.\n",
      "\n",
      "portable-gis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(teamengine) already exists.\n",
      "\n",
      "teamengine : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(mobilitydb) already exists.\n",
      "\n",
      "mobilitydb : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(bezitopo) already exists.\n",
      "\n",
      "bezitopo : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(eoxserver) already exists.\n",
      "\n",
      "eoxserver : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(nasaworldwind) already exists.\n",
      "\n",
      "nasaworldwind : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(firstdraftgis) already exists.\n",
      "\n",
      "firstdraftgis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geomesa) already exists.\n",
      "\n",
      "geomesa : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geopaparazzi) already exists.\n",
      "\n",
      "geopaparazzi : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotrellis) already exists.\n",
      "\n",
      "geotrellis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geowave) already exists.\n",
      "\n",
      "geowave : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(get-it) already exists.\n",
      "\n",
      "get-it : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(gisquick) already exists.\n",
      "\n",
      "gisquick : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(jtstopologysuite) already exists.\n",
      "\n",
      "jtstopologysuite : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(koop) already exists.\n",
      "\n",
      "koop : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(leaflet) already exists.\n",
      "\n",
      "leaflet : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(lerc) already exists.\n",
      "\n",
      "lerc : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(pdal) already exists.\n",
      "\n",
      "pdal : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(perfecttin) already exists.\n",
      "\n",
      "perfecttin : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(proj-jni) already exists.\n",
      "\n",
      "proj-jni : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(rasterframes) already exists.\n",
      "\n",
      "rasterframes : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(tegola) already exists.\n",
      "\n",
      "tegola : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(terraformer) already exists.\n",
      "\n",
      "terraformer : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotrellis) already exists.\n",
      "\n",
      "geotrellis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotrellis) already exists.\n",
      "\n",
      "geotrellis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotrellis) already exists.\n",
      "\n",
      "geotrellis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotrellis) already exists.\n",
      "\n",
      "geotrellis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geotrellis) already exists.\n",
      "\n",
      "geotrellis : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(pronto raster) already exists.\n",
      "\n",
      "pronto raster : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(geoserver-client-php) already exists.\n",
      "\n",
      "geoserver-client-php : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(wradlib) already exists.\n",
      "\n",
      "wradlib : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(oskari) already exists.\n",
      "\n",
      "oskari : Update successfully\n",
      "Exception: duplicate key value violates unique constraint \"project_pkey\"\n",
      "DETAIL:  Key (proj_id)=(qgis) already exists.\n",
      "\n",
      "qgis : Update successfully\n",
      "opendatacube : Insert successfully\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.6.13",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.6.13 64-bit ('3.6.13': pyenv)"
  },
  "interpreter": {
   "hash": "98373a4284e4a564d4f2df06473808b29be6466d133612bf3acd704edfd754d1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}